pyro/distributions/util.py

```diff
@@ -129,6 +129,38 @@ def detach(obj):
     return copy.deepcopy(obj, memo)
 
 
+class _DeepToMemo(dict):
+    def __init__(self, to_args, to_kwargs):
+        super().__init__()
+        self._to_args = to_args
+        self._to_kwargs = to_kwargs
+
+    def get(self, key, default=None):
+        result = super().get(key, default)
+        if result is default:
+            # Assume key is the id of another object, and look up that object.
+            old = ctypes.cast(key, ctypes.py_object).value
+            if isinstance(old, (torch.Tensor, torch.nn.Module)):
```

Comment:
Can you add a test exercising the `nn.Module` path? Also, will this work as expected on `PyroModule`s with `PyroSample`s and `PyroParam`s?

Reply:
Hmm I believe distribution attributes of modules will be bypassed, since we assume `Module` objects handle themselves via `torch.nn.Module.to()`. I'll have to think more about this...

Reply:
OK, I've added tests for `nn.Module`, `PyroModule`, `PyroParam`, and `PyroSample`. The only way I could get this to work was to perform the behavior out-of-place, by deep-copying the `nn.Module`. IMO this is the best we can do, given PyTorch's early design choices about behavior of `.to()` on tensors versus modules 🤷